src/dag_algo/mod.rs

```diff
@@ -637,3 +639,74 @@ pub fn collect_bicolor_runs(
 
     Ok(block_list)
 }
+
+/// Returns the transitive reduction of a directed acyclic graph
+///
+/// The transitive reduction of G = (V,E) is a graph G- = (V,E-)
+/// such that for all v,w in V there is an edge (v,w) in E- if and only if (v,w) is in E
+/// and there is no path from v to w in G with length greater than 1.
+///
+/// :param PyDiGraph graph: A directed acyclic graph
+///
+/// :returns: a directed acyclic graph representing the transitive reduction
+/// :rtype: PyDiGraph
+///
+/// :raises PyValueError: if graph is not a DAG
+
+#[pyfunction]
+#[pyo3(text_signature = "(graph, /)")]
+pub fn transitive_reduction(
+    graph: &digraph::PyDiGraph,
+    py: Python,
+) -> PyResult<digraph::PyDiGraph> {
+    let g = &graph.graph;
+    if !is_directed_acyclic_graph(graph) {
+        return Err(PyValueError::new_err(
+            "Directed Acyclic Graph required for transitive_reduction",
+        ));
+    }
+    let mut tr = StablePyGraph::<Directed>::new();
+    let mut descendants = HashMap::new();
+    let mut check_count = HashMap::new();
+    for (i, node) in g.node_indices().enumerate() {
+        tr.add_node(graph.get_node_data(i).unwrap().clone_ref(py));
+        check_count.insert(node, graph.in_degree(i));
+    }
+
+    for u in g.node_indices() {
+        let mut u_nbrs: HashSet<NodeIndex> = g.neighbors(u).collect();
+        for v in g.neighbors(u) {
+            if u_nbrs.contains(&v) {
+                if !descendants.contains_key(&v) {
+                    let dfs = dfs_edges(&g, Some(v));
+                    descendants.insert(v, dfs);
+                }
+                for desc in &descendants[&v] {
+                    u_nbrs.remove(&NodeIndex::new(desc.1));
+                }
+            }
+            *check_count.get_mut(&v).unwrap() -= 1;
+            if check_count[&v] == 0 {
+                descendants.remove(&v);
+            }
```

Comment:
If you want to avoid calling `StableGraph<...>.neighbors()` twice, you could instead make a copy of `u_nbrs` by calling `.clone()` on the loop. You could also pass a reference to said copy as the iterator in the loop, since in most cases you require a reference anyway.

```suggestion
        for v in &u_nbrs.clone() {
            if u_nbrs.contains(v) {
                if !descendants.contains_key(v) {
                    let dfs = dfs_edges(&g, Some(*v));
                    descendants.insert(*v, dfs);
                }
                for desc in &descendants[v] {
                    u_nbrs.remove(&NodeIndex::new(desc.1));
                }
            }
            *check_count.get_mut(v).unwrap() -= 1;
            if check_count[v] == 0 {
                descendants.remove(v);
            }
```

But it's all a question of choice. The code honestly works well regardless.

Reply:
In my mind I would try to avoid a call to clone `clone()` because the extra memory allocation and copying is typically slower. But it's totally worth benchmarking to see which ends up being faster.